# MerKurio: sequence extraction and annotation based on matching _k_-mers

</br>
<p align="center">    
    <a href="https://github.com/lschoenm/MerKurio/actions"><img alt="Build Status" src="https://img.shields.io/github/actions/workflow/status/lschoenm/MerKurio/tests.yml"></a>
    <a href="https://github.com/lschoenm/MerKurio/releases/latest"><img alt="GitHub Release" src="https://img.shields.io/github/v/release/lschoenm/MerKurio"></a>
    <a href="https://github.com/lschoenm/MerKurio/blob/master/LICENSE"><img alt="License" src="https://img.shields.io/badge/license-MIT-blue"></a>
</p>

MerKurio is a command line tool for extracting records from FASTA/FASTQ files based on _k_-mers, and for annotating and filtering aligned sequences in BAM/SAM format with _k_-mer tags.

It was developed to simplify downstream analysis of selected _k_-mers by tracing them back to their sequences in the original data.

MerKurio is designed to be user-friendly, flexible, fast, and memory efficient.

## Documentation

The full documentation can be found [here](https://lschoenm.github.io/MerKurio/).

A quick overview of the features and usage is provided below.

## Features

MerKurio provides two complementary subcommands:

- üîç **Extract**: Search FASTA/FASTQ data for _k_-mers and write records with matching _k_-mers to the terminal or a new file.
  - Supports paired-end reads (a hit in one read extracts the whole pair).
- üìë **Tag**: Annotate BAM/SAM alignments with _k_-mer tags and filter them based on matching _k_-mers.
  - Adds a two‚Äëletter tag (default `km`) with comma-separated matching k‚Äëmers (follows the [SAM format specification](https://samtools.github.io/hts-specs/SAMtags.pdf)).
  - Optionally keeps only reads containing at least one _k_‚Äëmer.
  - Multithreaded processing when working with BAM files.

Both commands share additional features:

- Records [detailed matching statistics](https://lschoenm.github.io/MerKurio/log.html) (positions of _k_-mer occurence, summary statistics, metadata).
  - Human readable output in plain text.
  - Structured JSON logs for easy machine parsing.
- Reads compressed input files (`.gz`, `.bz2`, `.xz`).
- Can seach for reverse complements or only canonical forms of _k_-mers.
- Case-insensitive search or conversion to lower-/uppercase.
- Inverse matching to keep only those records without matches.
- Query _k_-mers can be provided as command line arguments or in a file (FASTA or plain text).
- File types are inferred automatically.
- Record output can be suppressed to only record statistics.

## Example Workflow

For a quick and minimal example, follow the steps [described here](https://github.com/lschoenm/MerKurio/tree/master/example-minimal).

A more detailed practical example/tutorial can be found [in this repo](https://github.com/lschoenm/MerKurio/tree/master/example-workflow/).

Although the tool is designed to be flexible and can be used in a variety of ways, it was designed with the following workflow in mind:

1. Have a list of _k_-mers of interest.
2. Use **MerKurio** to extract the FASTA/FASTQ records containing these _k_-mers (and their reverse complements) from the original (paired-end) records.

If sequencing reads were extracted:

3. Align the extracted reads to a reference genome.
4. Use **MerKurio** to tag the aligned records in a SAM/BAM file with the significant _k_-mers they contain (and optionally filtering it).
5. Analyzing the matching statistics generated by MerKurio.
6. Interpret the results: e. g., by visualizing the _k_-mers that are frequently tagged in the reads that align to a particular region of the genome.

## Usage

Run `merkurio` or one of its subcommands with the `--help` flag to see the available options and subcommands.

### The `extract` Subcommand

Running `merkurio extract` will extract records from a sequence file (FASTA or FASTQ, format is inferred automatically) based on a list of query sequences (_k_-mers). The query sequences can be provided in a file or as a list of strings on the command line. Reverse complements can be included in the search. The extracted records are written to stdout or to a new file in the same format as the input file. The tool tries to select the most efficient algorithm for the given query sequences.

Detailed match statistics are written to stdout or to a file if specified, showing which records got hit by sequences along with a zero-based position. Matching statistics can also be saved in JSON format for easier parsing by other programs.

An example usage of the `extract` subcommand to extract records from a FASTA file based on a list of _k_-mers and their reverse complements is shown below; logging information is written to stdout:

```bash
merkurio extract -i input.fasta -o output.fasta -f query_kmers.txt -r -l
```

Another example where paired-end reads are extracted if they contain the sequence ACGT or TGCA; the extracted records are written to stdout and logging information is written to a file called `log.txt` (the `-i` and `-1` flags can be used interchangeably):

```bash
merkurio extract -1 input_R1.fastq -2 input_R2.fastq -o output -s ACGT TGCA -l log.txt
```

### The `tag` Subcommand

Running `merkurio tag` will tag aligned sequences in a BAM/SAM file with _k_-mers. If a record contains one or more of the _k_-mers, it is annotated with a tag ("km" by default; must be exactly two characters long) and the respective _k_-mers. Multithreading is supported for BAM files. Optionally, keep only records which are matching at least one _k_-mer.

Detailed match statistics are written to stdout or to a file if specified, showing which records got hit by sequences along with a zero-based position. Matching statistics can also be saved in JSON format for easier parsing. Matching records output can be suppressed if one is only interested in the matching statistics.

An example usage of the `tag` subcommand to tag a BAM file with the _k_-mers in the file `query_kmers.fasta`, with SAM output:

```bash
merkurio tag -i input.bam -o output.sam -f query_kmers.fasta
```

Another example where the _k_-mers are provided on the command line, and the search is also performed for their reverse complements. The tag is set to "MK". BAM file processing is done with 4 threads:

```bash
merkurio tag -i input.bam -o output.bam -s ACGT TGCA -r -p 4 -t MK
```

## Installation

Different installation methods are available for **MerKurio**, depending on whether the user has `cargo`/Rust installed and wants to build from source, or only download the executable.

You can test if the installation was successful by executing MerKurio:

```bash
# In case you downloaded the standalone executable
./path/to/merkurio --help

# If MerKurio is added to PATH
merkurio --help
```

### Precompiled Binaries

The most straightforward way with no external dependencies.

Precompiled binaries for Linux, Windows, and MacOS are available on the [releases page](https://github.com/lschoenm/MerKurio/releases). Download the appropriate binary for your system, decompress and untar it. Optionally, add it to your PATH. The "musl" Linux version can be slower on some systems than the "gnu" version, but should be compatible with a wider range of systems.

To decompress the gzipped tarball, run the following command:

```bash
tar -xzf path/to/release.tar.gz
```

Then, add user permissions (`chmod u+x path/to/merkurio`) to be able to execute the program from this path.

### Install a Release using `cargo`

Alternatively, you can install a released version of **MerKurio** from source using `cargo`. This requires you to have Rust installed, [which is described here](https://doc.rust-lang.org/cargo/getting-started/installation.html). To install a specific version of MerKurio, run the following command (replace `X.X.X` with the version number):

```bash
cargo install --git https://github.com/lschoenm/MerKurio --tag vX.X.X
```

#### Install from Source

To install from source, download the source code from the [releases page](https://github.com/lschoenm/MerKurio/releases):

```bash
cargo install --path path/to/source/
```

Or clone the repo to get the latest version, decompress and untar it, and install MerKurio using `cargo`. This will automatically add it to your path:

```bash
git clone https://github.com/lschoenm/MerKurio
cd MerKurio
cargo install --path .
```

#### Build from Source

In order to build the program from source without having it added to your PATH, clone the repository and run the following command in the root directory. The binary will be located in the `target/release` directory:

```bash
git clone https://github.com/lschoenm/MerKurio
cd MerKurio
cargo build --release
```
