# Example Workflow

A practical example/tutorial can be found [in the repository](https://github.com/lschoenm/MerKurio/tree/master/example-workflow/) or on the next page. 

Although the tool is designed to be flexible and can be used in a variety of ways, it was designed with the following workflow in mind:

1. Have a list of _k_-mers of interest. 
2. Use **MerKurio** to extract the FASTA/FASTQ records containing these _k_-mers (and their reverse complements) from the original (paired-end) records. 

If sequencing reads were extracted: 

3. Align the extracted reads to a reference genome. 
4. Use **MerKurio** to tag the aligned records in a SAM/BAM file with the significant _k_-mers they contain (and optionally filtering it). 
5. Analyzing the matching statistics generated by MerKurio. 
6. Interpret the results: e. g., by visualizing the _k_-mers that are frequently tagged in the reads that align to a particular region of the genome. 